src/IronRure/RegexSet.cs

```diff
@@ -3,131 +3,150 @@
-        /// </summary>
-        public bool IsMatch(byte[] haystack)
-        {
-            return RureFfi.rure_set_is_match(_handle, haystack,
-                                             new UIntPtr((uint)haystack.Length),
-                                             UIntPtr.Zero);
-        }
+        foreach (GCHandle handle in patByteHandles)
+            handle.Free();
 
-        /// <summary>
-        ///   Matches - Retrieve information about which patterns in the set match.
-        /// </summary>
-        public SetMatch Matches(string haystack) =>
-            Matches(Encoding.UTF8.GetBytes(haystack));
+        if (compiled.IsInvalid)
+            throw new RegexCompilationException(err.Message);
 
-        /// <summary>
-        ///   Matches - Retrieve information abut which patterns in the set match.
-        /// </summary>
-        public SetMatch Matches(byte[] haystack)
-        {
-            var results = new bool[_arity];
-            var overall = RureFfi.rure_set_matches(_handle,
-                                                   haystack,
-                                                   new UIntPtr((uint)haystack.Length),
-                                                   UIntPtr.Zero,
-                                                   results);
-            return new SetMatch(overall, results);
-        }
+        return compiled;
+    }
+
+    /// <summary>
+    ///   Is Match - Checks if any of the patterns in the set match.
+    /// </summary>
+    public bool IsMatch(string haystack) =>
+        IsMatch(Encoding.UTF8.GetBytes(haystack));
 
-        /// <inheritdoc />
-        public void Dispose()
+    /// <summary>
+    ///   Is match - Check if any of the patterns in the set match.
+    /// </summary>
+    public bool IsMatch(byte[] haystack)
+    {
+        return RureFfi.rure_set_is_match(_handle, haystack,
+                                         new UIntPtr((uint)haystack.Length),
+                                         UIntPtr.Zero);
+    }
+
+    /// <summary>
+    ///   Matches - Retrieve information about which patterns in the set match.
+    /// </summary>
+    public SetMatch Matches(string haystack) =>
+        Matches(Encoding.UTF8.GetBytes(haystack));
+
+    /// <summary>
+    ///   Matches - Retrieve information abut which patterns in the set match.
+    /// </summary>
+    public SetMatch Matches(byte[] haystack)
+    {
+        bool[] results = new bool[_arity];
+        bool overall = RureFfi.rure_set_matches(_handle,
+                                               haystack,
+                                               new UIntPtr((uint)haystack.Length),
+                                               UIntPtr.Zero,
+                                               results);
+        return new SetMatch(overall, results);
+    }
+
+    /// <summary>
+    ///   Finalizer to ensure resources are released.
+    /// </summary>
+    ~RegexSet()
```

Comment:
I don't think we directly have any unmanaged resources here? WHy the finaliser?

Reply:
I think the old dispose impl was correct.

Reply:
Reverted

Reply:
Or rather, improved rather than reverted.